Surface agent modules health from cilium status command

derailed opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

<!-- Description of change -->

- Add hive module health injection to support for module health reporting (pulled from @joamaki prototype)
- Add swagger health apis
- Surface agent modules health via agent CLI

> There are dependencies here on these 2 PRs: https://github.com/cilium/cilium/pull/25725, https://github.com/cilium/cilium/pull/25662 

Signed-off-by: Fernand Galiana <fernand.galiana@isovalent.com>

```release-note
Augments `cilium status` CLI to report on agent modules health status.
```

derailed committed:
Fake out node manager module to report health

- Add fake health report for testing

> NOTE: This will need to be updated based on PR
(https://github.com/cilium/cilium/pull/25725).

Signed-off-by: Fernand Galiana <fernand.galiana@isovalent.com>

derailed committed:
Add modules health report

The agent status CLI command needs to be able to output modules health information.
As we begin implementing modules health status reporting, we need a new
health section in the CLI output to present users with modules status
specifics.

- Surface agent modules health on the `cilium status` command

Signed-off-by: Fernand Galiana <fernand.galiana@isovalent.com>

tommyp1ckles commented on pkg/node/manager/manager.go:
```diff
@@ -294,11 +299,13 @@ func (m *manager) backgroundSync(ctx context.Context) error {
 
 			m.metricDatapathValidations.Inc()
 		}
+		m.healthReporter.OK("All good here! Nothing to see")
```
We can drop this code from this PR, I trust that this is all working and will pick up any modules reporting a status in future PRs.

derailed replied:
Indeed... Thanks Tom!

tommyp1ckles commented on pkg/datapath/linux/probes/managed_neighbors.go:
```diff
@@ -83,6 +83,9 @@ func haveManagedNeighbors() (outer error) {
 		return fmt.Errorf("failed to add dummy veth: %w", err)
 	}
 
+	if err := netlink.LinkAdd(veth); err != nil {
```
Test code - should probably remove?

tommyp1ckles commented on api/v1/openapi.yaml:
```diff
@@ -1,33 +1,33 @@
 ---
-swagger: '2.0'
+swagger: "2.0"
```
Did your editor making these changes? we should avoid committing auto-linting changes, especially if they're not codified in the codebase.

derailed replied:
Rats! it did - will update

nathanjsweet replied:
This isn't resolved. There are still a lot of spurious changes; please revert all of these redundant changes.

derailed replied:
Rats! good catch. Tx Nate! tabs and valuable EOL spaces now restored to their former glory.

zacharysarah commented on api/v1/openapi.yaml:
```diff
@@ -1202,6 +1221,33 @@ parameters:
 
 definitions:
+  ModulesHealth:
+    description: Reports agent's modules health status
```
Clarity nit
```suggestion
    description: Reports health status of agent's modules
```

derailed replied:
@zacharysarah Good catch. Thank you!

zacharysarah reviewed (approved):
One nit to make it easier to differentiate ModulesHealth from ModuleHealth, otherwise LGTM from a docs perspective. Approving with the understanding that the change is required prior to merge.

thorn3r commented on pkg/node/manager/manager.go:
```diff
@@ -294,6 +299,7 @@ func (m *manager) backgroundSync(ctx context.Context) error {
 			m.mutex.RUnlock()
 			m.Iter(func(nh datapath.NodeHandler) {
 				nh.NodeValidateImplementation(entry.node)
+
```
tiny nit: whitespace change here

thorn3r commented on pkg/datapath/linux/probes/managed_neighbors.go:
```diff
@@ -83,9 +83,6 @@ func haveManagedNeighbors() (outer error) {
 
-	if err := netlink.LinkAdd(veth); err != nil {
-		return fmt.Errorf("failed to add dummy veth: %w", err)
-	}
```
Since this was for testing can we drop the change from both commits? I imagine this could cause problems during a bisect

derailed replied:
@thorn3r Thank you for the review! I'll axe it

derailed committed:
Add agent health endpoint

We need a new endpoint to collect modules health status via the agent api.
This will be used initialy by the `cilium status` command to hydrate a new
section listing out all instrumented modules health.

- Update swagger config to surface health endpoint

Signed-off-by: Fernand Galiana <fernand.galiana@isovalent.com>

rolinh reviewed (approved):
Vendor changes lgtm

thorn3r reviewed (approved):
lgtm! :+1:

qmonnet reviewed (approved):
Looks good from my side

christarazi commented on daemon/cmd/daemon.go:
```diff
@@ -216,7 +217,9 @@ type Daemon struct {
 
-	l2announcer *l2announcer.L2Announcer
+	l2announcer    *l2announcer.L2Announcer
+	healthProvider cell.Health
```
Let's add a newline between these to represent the logical separation of these two fields

derailed replied:
@christarazi Thanks for the review. Makes sense!

christarazi commented on daemon/cmd/vitals.go:
```diff
@@ -0,0 +1,62 @@
+	"time"
+
+	"github.com/go-openapi/runtime/middleware"
+	"k8s.io/apimachinery/pkg/util/duration"
```
Can you squash the vendor commit into this commit since it's being used here and it's fairly small? This makes bisectability better as the new dependency is added with the implementation.

derailed replied:
I am a bit confused with the heuristics to track dependencies. On previous commit I was ask to specifically keep dependencies update in separate commit. But here since it's small it's ok to roll it in?

christarazi replied:
Ah I see. I'll share more on why I made this comment: as I reviewed commit by commit, it was difficult to tell why the "Update dependencies" commit was necessary because there was no context in the msg for why it's being included. Later, I noticed the new package was included in used in the "Add modules health report" commit. 

Because the diff of the new dependency was so small, I probably wouldn't have reviewed in if it was squashed into the latter commit. The package was already included indirectly as a dependency of Cilium and the main diff is that it's now a direct dependency (that's why only the `vendor/` changed). Given this, I'd say it's a very minor change that doesn't require a separate review (by being in its own commit). If we were including a new module (requiring changes to `go.mod`) then I'd say a separate commit makes sense. 

However, if the new module is small enough anyway, then you can get away with it being bundled. I'd say the main heuristic is what makes the most sense to reviewers. Should they review the new dependency? If so, make it separate with context in the commit msg. If the dependency is not controversial*, probably just include it with the implementation commit.

*: I'm hand waving a bit here for expediency purposes, but I realize we probably need to document better what are our expectations around vendor changes and commits.

derailed replied:
Thanks for the clarifications and explanations Chris!
 I thought since part of the review council are folks specifically in charge of checking for dependencies updates, it made sense for it to be in a separate commit so it can ease there review no matter how small. Tho no go.mod was changed /vendor did. One the other hand and as you pointed at, I makes sense for it to be part of the commit that warrants it.

christarazi commented on daemon/cmd/vitals.go:
```diff
@@ -0,0 +1,62 @@
+	"github.com/cilium/cilium/pkg/hive/cell"
+)
+
+const naValue = "n/a"
```
Nit: No need for this const if it's just used once

christarazi commented on cilium/cmd/status.go:
```diff
@@ -107,7 +107,7 @@ func statusDaemon() {
 			healthPkg.GetAndFormatModulesHealth(w, client.Daemon, allHealth)
 		} else {
-			fmt.Fprint(w, "Cluster health:\t\tProbe disabled\n")
+			fmt.Fprint(w, "Cluster Health:\t\tProbe disabled\n")
```
While I appreciate the intent of these changes, I don't particularly think they should be part of this PR and merged late into the release cycle. My short term concern is that we might have scripts somewhere relying on this. Longer term, without a style guide to point to, these changes are arbitrary and "subjective". Someone later down the line might come along and propose the effective inverse of these changes, :shrug:.

My proposal is to drop this commit for now.

Happy to discuss and pull in the right folks if we think a style guide for our CLI outputs to be more consistent / following a particular pattern is something we should pursue.

derailed replied:
Makes sense. I'll axe it. We have a bit of a `Frankenstein` theme going at the moment regarding the output with a strong majority to capitalize, hence this effort to normalize the output for consistency. 
That said, If scripts are scraping human readable console format while we have options to use structured data via yaml or json, we have other issues.

christarazi commented on daemon/cmd/daemon.go:
```diff
@@ -216,7 +217,9 @@ type Daemon struct {
 
-	l2announcer *l2announcer.L2Announcer
+	l2announcer    *l2announcer.L2Announcer
+	healthProvider cell.Health
+	healthReporter cell.HealthReporter
```
Speaking of logical separations, what do you think of combining the reporter into the provider? Do we really need two separate cells? At a surface level, it seems reasonable that a health provider has a "reporter" component.

derailed replied:
I think SCP usage here make sense as these are managing separate concerns.

christarazi reviewed (changes requested):
Nice work!

I have a few comments regarding the PR. My other comment is that I'd like to see some motivation written in each commit msg describing the "why", rather than just the "what". This allows us to preserve the original intent in the history for better bisect and context sharing in general for code readers.

asauber commented on api/v1/client/daemon/daemon_client.go:
```diff
@@ -211,6 +213,46 @@ func (a *Client) GetDebuginfo(params *GetDebuginfoParams, opts ...ClientOption)
+	}
+	// unexpected success response
+	// safeguard: normally, absent a default response, unknown success responses return an error above: so this is a codegen issue
+	msg := fmt.Sprintf("unexpected success response for GetHealth: API contract not enforced by server. Client expected to get an error, but got: %T", result)
```
Using `%#v` rather than `%T` here would be useful, as it would include both the type and value rather than just the type

derailed replied:
Not sure if this would be that useful here given we are expecting a specific type that somehow did not come across the wire. Imho this is a clearer error that dumping the type and entire resp content which would not be that actionable?

asauber replied:
You're right, I agree that %T is better.

asauber commented on api/v1/client/daemon/get_health_parameters.go:
```diff
@@ -0,0 +1,159 @@
+}
+
+/*
+GetHealthParams contains all the parameters to send to the API endpoint
```
Comment formatting here looks like it was mangled.

asauber commented on api/v1/client/daemon/get_health_parameters.go:
```diff
@@ -0,0 +1,159 @@
+
+	   Brief will return a brief representation of the Cilium status.
+
+	*/
```
```suggestion
	// Brief is a brief representation of the Cilium status.
```

I'm aware that this is autogenerated, maybe there's a way to hint that this be formatted better.

derailed replied:
Good catch!

asauber commented on api/v1/client/daemon/get_health_parameters.go:
```diff
@@ -0,0 +1,159 @@
+	var res []error
+
+	if o.Brief != nil {
+
```
nit: extra newline

derailed replied:
This is generated code

aanm reviewed (commented):
It's not clear to me why we can't extend the current `GetHealthz` API endpoint for this purpose given that `GetHealth` seems to have the same parameters as `GetHealthz`

asauber commented on api/v1/models/modules_health.go:
```diff
@@ -0,0 +1,119 @@
+			continue
+		}
+
+		if m.Modules[i] != nil {
+			if err := m.Modules[i].Validate(formats); err != nil {
+				if ve, ok := err.(*errors.Validation); ok {
+					return ve.ValidateName("modules" + "." + strconv.Itoa(i))
+				} else if ce, ok := err.(*errors.CompositeError); ok {
+					return ce.ValidateName("modules" + "." + strconv.Itoa(i))
+				}
+				return err
+			}
+		}
```
```suggestion
		if m.Modules[i] == nil {
			continue
		}
		
		if err := m.Modules[i].Validate(formats); err != nil {
			if ve, ok := err.(*errors.Validation); ok {
				return ve.ValidateName("modules" + "." + strconv.Itoa(i))
			} else if ce, ok := err.(*errors.CompositeError); ok {
				return ce.ValidateName("modules" + "." + strconv.Itoa(i))
			}
			return err
		}
```

asauber replied:
I'm aware that this is auto-generated. Kind of a shame that the code is unnecessarily nested.

derailed replied:
Right! short of PRing their repo not sure if there is much we can do here??

asauber replied:
Yep

asauber commented on api/v1/models/modules_health.go:
```diff
@@ -0,0 +1,119 @@
+
+	for i := 0; i < len(m.Modules); i++ {
+
+		if m.Modules[i] != nil {
+			if err := m.Modules[i].ContextValidate(ctx, formats); err != nil {
+				if ve, ok := err.(*errors.Validation); ok {
+					return ve.ValidateName("modules" + "." + strconv.Itoa(i))
+				} else if ce, ok := err.(*errors.CompositeError); ok {
+					return ce.ValidateName("modules" + "." + strconv.Itoa(i))
+				}
+				return err
+			}
+		}
+
+	}
```
```suggestion
		if m.Modules[i] == nil {
			continue
		}
		
		if err := m.Modules[i].ContextValidate(formats); err != nil {
			if ve, ok := err.(*errors.Validation); ok {
				return ve.ValidateName("modules" + "." + strconv.Itoa(i))
			} else if ce, ok := err.(*errors.CompositeError); ok {
				return ce.ValidateName("modules" + "." + strconv.Itoa(i))
			}
			return err
		}
```

asauber commented on api/v1/server/restapi/daemon/get_health_parameters.go:
```diff
@@ -0,0 +1,82 @@
+	// HTTP Request Object
+	HTTPRequest *http.Request `json:"-"`
+
+	/*Brief will return a brief representation of the Cilium status.
```
```suggestion
	/*Brief is a brief representation of the Cilium status.
```

derailed replied:
Tx Andrew!

asauber commented on api/v1/server/restapi/daemon/get_health_parameters.go:
```diff
@@ -0,0 +1,82 @@
+		raw = rawData[len(rawData)-1]
+	}
+
+	// Required: false
```
I'm aware that this is autogenerated, do you know what this comment is in reference to?

tommyp1ckles replied:
I'm guessing it's this: https://github.com/cilium/cilium/pull/25714/files#diff-508a09ed0e66f45b2f1c78e3650eedf4de551b63150c6dac457298af46f9152bR67 ?

asauber commented on pkg/health/client/modules_test.go:
```diff
@@ -0,0 +1,92 @@
+	}
+}
+
+// Helpers...
```
The `...` makes me nervous 😅

derailed replied:
Not sure I get the reference??

asauber reviewed (approved):
Overall looks good, some nits and formatting questions.
